### PR TITLE
Development to master

### PR DESCRIPTION
**Pull request by SamZand**
Willem, Arsalan,
This module runs on the Raspberry Pi, can take pictures (using the raspistill command) and videos (using raspivid) and can produce a base64 encoded string. This is done by converting the redirected output from the raspivid command to base64.
This base64 string is supplied to a lambda function, which can then use it for its own purposes.

### CODE REVIEW

File: src/base64_encoder.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+std::string Camera::Base64Encoder::encode() {
+
+    std::vector<unsigned char> &data = getByteStream();
+
+    std::string result;
+    int i = 0, j = 0;
+
+    unsigned char array3[3];
```
Comment (ghost) [inline src/base64_encoder.cpp]: Please only use the c-style array when no other option is available.

File: src/base64_encoder.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    return result;
+}
+
+std::string Camera::Base64Encoder::encode(std::vector<unsigned char> &data) {
+
+    std::string result;
+    int i = 0, j = 0;
+
+    unsigned char array3[3];
```
Comment (ghost) [inline src/base64_encoder.cpp]: Please only use the c-style array when no other option is available.

File: src/base64_encoder.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    return buffer;
+}
+
+void Camera::Base64Encoder::setFilePath(std::string &newFilePath) {
+    filePath = newFilePath;
+}
+
+std::string Camera::Base64Encoder::encode() {
+
+    std::vector<unsigned char> &data = getByteStream();
```
Comment (ghost) [inline src/base64_encoder.cpp:30]: This function is exactly the same as the other encode() function except for this line. Why not call that function after you get the data with this line.

File: src/pi_cam.hpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * @brief     Header for Pi camera
+ * @author    Willem de Groot, Arsalan Anwari
+ * @license   See LICENSE
+ */
+#ifndef PICAM_HPP
+#define PICAM_HPP
+
+#include "base64_encoder.hpp"
+#include "camera.hpp"
+#include <array>
```
Comment (ghost) [inline src/pi_cam.hpp]: Unused include.

File: src/vector_2d.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    int getY() const;
+    /**
+     * @brief Set the y value of the vector.
+     *
+     * @param[in] inputY  The desired y value.
+     */
+    void setY(int inputY);
+
+    std::string print();
```
Comment (ghost) [inline src/vector_2d.hpp]: Undocumented function.

File: src/vector_2d.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    // friend std::ostream& operator<<(std::ostream& os, const Vector2D & rhs);
+
+    bool operator==(const Vector2D &rhs) const;
```
Comment (ghost) [inline src/vector_2d.hpp]: Undocumented function.

**Review by ghost — CHANGES_REQUESTED** [review]
Some things have to change, but nothing that isn't easily fixable.

File: src/pi_cam.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+std::string Camera::PiCam::getSettings() {
+    std::stringstream result;
+    result << " -fps " << fps << " -w " << resolution.getX() << " -h " << resolution.getY();
+    return result.str();
+}
+
+void Camera::PiCam::printSettings() {
+    std::cout << "Resolution: " << getResolution().print() << "\n"
```
Comment (JasperSmienk) [inline src/pi_cam.cpp]: std::endl is preferred over \n as it flushes, unless there is a performance reason not to use it.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
Looks good now. Merging.

**Pull request merged**